src/Model/Operation/ProductSyncHandler.php

```diff
@@ -274,6 +286,28 @@ private function handleConfiguratorGroups(SalesChannelProductEntity $product): a
         return $mainProducts;
     }
 
+    private function handleFirstActiveVariant(ProductEntity $product): ProductEntity
+    {
+        $mainProduct = null;
+        $variants = new ProductCollection([$product]);
+
+        foreach ($product->getChildren() as $child) {
+            if ($child === $product->getChildren()->filterByProperty('active', true)->first()) {
+                $mainProduct = $child;
+            } else {
+                $variants->add($child);
+            }
+        }
+
+        if (!$mainProduct) {
+            return $product;
+        }
+
+        $mainProduct->setChildren($variants);
+
+        return $mainProduct;
```

Comment:
You are looping over all children within each run of the children loop. Something like this should be enough, which is a combination from both of your solutions, but has only one loop in total (and also none of my suggested functions 😅):

This will need some logic change where the function is called

```suggestion
        $mainProduct = null;
        $variants = new ProductCollection([$product]);

        foreach ($product->getChildren() as $child) {
            if ($child->getActive() && !$mainProduct) {
                $mainProduct = $child;
            } else {
                $variants->add($child);
            }
        }

        if ($mainProduct) {
            $mainProduct->setChildren($variants);
        }

        return $mainProduct;
```